### PR TITLE
[11.x] Adds conditional to routes

### PR DESCRIPTION
**Pull request by Boorinio**
This pull request introduces the Conditionable trait to the Laravel Route class, allowing developers to add conditional logic when defining routes. Consider the following:

```
Route::middleware('shop')
    ->domain('{shop}.domain.com')
    ->when(App::isProduction(), function ($route) {
        $route->whereIn('shop', app(ShopService::class)->getShopSlugs());   
    });
```
### No Breaking Changes
This addition does not alter any existing functionality of the Route class. The Conditionable trait is introduced as an optional feature that will only be applied when explicitly used by developers. Existing routes will continue to work as expected without any modifications.

P.S. tests included. 

Thanks for your time

### CODE REVIEW

**Comment by Boorinio** [comment]
@taylorotwell Thank you, much appreciated!!

**Comment by Eyad-Bereh** [comment]
Can you please explain how will this feature be useful? Any specific use cases you were thinking of?

**Comment by edalzell** [comment]
> Can you please explain how will this feature be useful? Any specific use cases you were thinking of?

In my case I have a route that's only used in testing, this is perfect for that.

**Comment by Eyad-Bereh** [comment]
> > Can you please explain how will this feature be useful? Any specific use cases you were thinking of?
> 
> In my case I have a route that's only used in testing, this is perfect for that.

If you need specific routes for testing, shouldn't you just define them inside the test?
I'm talking about something like (https://github.com/laravel/framework/blob/5a880cc8a1d424c0958c80a199a07bf3ea7e9850/tests/Routing/RoutingRouteTest.php#L54) instead of hardcoding them inside your project code?

File: tests/Routing/RoutingRouteTest.php [hunk 1]
```diff
@@ -943,6 +943,26 @@ public function testWherePatternsProperlyFilter()
...
+                'one',
+                'two',
+            ]);
+        });
+
+        $request = Request::create('test.awesome.test', 'GET');
+        $this->assertFalse($route->matches($request));
+
+        $request = Request::create('one.awesome.test', 'GET');
+        $this->assertTrue($route->matches($request));
```
Comment (shaedrich) [inline tests/Routing/RoutingRouteTest.php:965]: This test doesn't actually test if `when()` works but if `matches()` works—this is counterintuitive to have in this PR

**Pull request merged**